src/blockMonitor/blockCache.ts

```diff
@@ -99,110 +92,142 @@ export class BlockCache<TBlock extends IBlockStub> implements ReadOnlyBlockCache
      * Returns true it `block` can be added to the cache, that is if either:
      *   - it is the first block ever seen, or
      *   - its parent is already in the cache, or
-     *   - it is at depth least `this.maxDepth`.
+     *   - it is at exactly `this.maxDepth`.
+     * If not, it can only be added to the pending.
      * @param block
      */
     public canAddBlock(block: Readonly<TBlock>): boolean {
-        return this.isEmpty || this.hasBlock(block.parentHash) || block.number <= this.minHeight;
+        return this.isEmpty || this.hasBlock(block.parentHash) || block.number === this.minHeight;
     }
 
-    /**
-     * `canAddBlock` might return true for blocks that should actually not be added.
-     * Here we check that the block is not actually already added, and it is not below a height
-     * that would be pruned immediately.
-     */
-    private shouldAddBlock(block: Readonly<TBlock>) {
-        if (this.blockStubsByHash.has(block.hash)) {
-            // block already in memory
-            return false;
+    // Processes all pending blocks at the given height, moving them to blocksByHeight if they are now complete.
+    // If so, add them and repeat with the next height (as some blocks might now have become complete)
+    private processPending(height: number) {
+        const blockHashesAtHeight = this.blockHashesByHeight.get(height) || new Set();
+        const blockHashesToAdd = [...blockHashesAtHeight].filter(h => this.pendingBlocksByHash.has(h));
+
+        for (const blockHash of blockHashesToAdd) {
+            const block = this.pendingBlocksByHash.get(blockHash)!;
+
+            // Remove block from pendingBlocksByHash, add to blocksByHash
+            this.blocksByHash.set(blockHash, block);
+            this.pendingBlocksByHash.delete(blockHash);
+
+            // Might need to update the maximum height
+            this.updateMaxHeightAndPrune(block.number);
         }
 
-        if (block.number < this.minHeight) {
-            // block too deep
-            return false;
+        if (blockHashesToAdd.length > 0) {
+            this.processPending(height + 1);
+        }
+    }
+
+    // If minHeight is increased after adding some blocks, some previously pending blocks should now be moved into blocksByHash.
+    // Since the process itself could (in rare circumstances) also increase minHeight, we check if this is the case and repeat the cycle.
+    private processPendingBlocksAtMinHeight() {
+        let prevMinHeight: number;
+        do {
+            prevMinHeight = this.minHeight;
+            this.processPending(this.minHeight);
```

Comment:
It can never be that `maxHeight === minHeight`, except at startup.

This is processing pending blocks that might have become ready to add ("complete") because minHeight increased (==> because `maxHeight` did).

Calling `this.processPending(this.minHeight);` could increase `maxHeight` if there is a chain of pending blocks from height `minHeight` until `maxHeight + 1` (or more). In this case `minHeight` also increases, so we repeat the process. It seems very unlikely to happen (a very long separate chain!), but we can handle it by just repeating that until no change in minHeight happens after `processPending`.

Reply:
what i thought could happen cant, because of a guard in update max height
```
if (this.mMaxHeight < newHeight) {
```